### PR TITLE
Remove legacy support for the magic `log`/`progress_bar` keys in dict returns

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

See title

Also removes some unnecessary dev_debuggers

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: tests/trainer/data_flow/test_train_loop_flow_dict.py [hunk 1]
```diff
@@ -24,7 +22,6 @@
 from tests.helpers.deterministic_model import DeterministicModel
 
 
-@mock.patch.dict(os.environ, {"PL_DEV_DEBUG": "1"})
```
Comment (Borda) [inline tests/trainer/data_flow/test_train_loop_flow_dict.py:27]: maybe leave this `PL_DEV_DEBUG` for a separate PR so this is focus on one thing only

**Review by Borda — APPROVED** [review]
Finally :D

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -361,6 +318,8 @@ def _process_training_step_output_1_0(self, training_step_output, split_batch):
         if isinstance(training_step_output, dict):
             loss = training_step_output.pop("loss", None)
             hiddens = training_step_output.pop("hiddens", None)
+            if hiddens is not None:
+                hiddens = hiddens.detach()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:322]: We are detaching here because of truncated backpropagation through time, right?
So was this simply missing before?

File: tests/trainer/logging_/test_logger_connector.py [hunk 3]
```diff
@@ -208,12 +208,10 @@ def training_step_end(self, *_):
 
 
 @pytest.mark.parametrize('num_dataloaders', [1, 2])
-def test__logger_connector__epoch_result_store__test_multi_dataloaders(tmpdir, monkeypatch, num_dataloaders):
+def test__logger_connector__epoch_result_store__test_multi_dataloaders(tmpdir, num_dataloaders):
```
Comment (awaelchli) [inline tests/trainer/logging_/test_logger_connector.py:211]: there are almost more underscores than characters in this test name xD

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:322]: Yes. I moved it to the wrong place in #6016 so this fixes that unreleased bug
